src/base/EIP712.sol

```diff
@@ -0,0 +1,44 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.20;
+
+import {IEIP712} from "../interfaces/IEIP712.sol";
+
+/// @notice Generic EIP712 implementation
+/// @dev Maintains cross-chain replay protection in the event of a fork
+/// @dev Reference: https://github.com/Uniswap/permit2/blob/main/src/EIP712.sol
+/// @dev Reference: https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/cryptography/EIP712.sol
+contract EIP712 is IEIP712 {
+    // Cache the domain separator as an immutable value, but also store the chain id that it
+    // corresponds to, in order to invalidate the cached domain separator if the chain id changes.
+    bytes32 private immutable _CACHED_DOMAIN_SEPARATOR;
+    uint256 private immutable _CACHED_CHAIN_ID;
+    bytes32 private immutable _HASHED_NAME;
+    bytes32 private immutable _HASHED_VERSION;
+
+    bytes32 private constant _TYPE_HASH =
+        keccak256("EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)");
```

Comment:
do we rlly need version? whats the gas with no version?

Reply:
increased first-cold-permit by 12 gas but saved 12 gas if its warm already